open_spiel/games/phantom_go.cc

```diff
@@ -29,84 +29,178 @@ namespace {
+    std::unique_ptr<PhantomGoState>
+        newState = std::make_unique<PhantomGoState>(down_cast<PhantomGoState>(*newGame->NewInitialState()));
+
+    std::array<GoColor, kMaxBoardSize * kMaxBoardSize> infoState = board_.GetObservationByID(player_id);
+    std::array<int, 2> stoneCount = board_.getStoneCount();
+
+    std::array<std::vector<int>, 2> stones;
+    std::vector<Action> enemyActions;
+    std::vector<bool> enemyActionVisibility;
+    std::vector<int> enemyActionNumber;
+
+    auto opp_payer_id = (uint8_t) OppColor((GoColor) player_id);
+
+    //Find and store all stones which are in the last move on board
+    for (int i = 0; i < boardSize * boardSize; i++) {
+        if (infoState[i] != GoColor::kEmpty) {
+            stones[(uint8_t) infoState[i]].push_back(i);
+        }
+    }
+
+    if(player_id == (uint8_t)GoColor::kWhite)
+    {
+        newState->ApplyAction(VirtualActionToAction(kVirtualPass, boardSize));
+    }
+
+    for(long action : stones[player_id]) // Fill the board with stones of player we want to resample for
+    {
+        newState->ApplyAction(action);
+        newState->ApplyAction(VirtualActionToAction(kVirtualPass, boardSize));
+    }
+
+    if(!newState->history_.empty())
+    {
+        newState->UndoAction(-1, -1);
+    }
+
+
+    auto opp_player_id = (uint8_t)OppColor((GoColor)player_id);
+    for(long action : stones[opp_player_id])
+    {
+        newState->ApplyAction(action);
+        newState->ApplyAction(action);
+        newState->ApplyAction(VirtualActionToAction(kVirtualPass, boardSize));
+    }
+
+    for(int i = 0; i < stoneCount[opp_player_id] - stones[opp_player_id].size(); i++) {
+        std::vector<Action> actions = newState->LegalActions();
+        std::shuffle(actions.begin(), actions.end(), std::mt19937(std::random_device()()));
+        std::array<int, 2> currStoneCount = newState->board_.getStoneCount();
+        currStoneCount[opp_player_id]++;
+        std::vector<int> vec = stones[opp_player_id];
+
+        for (long action: actions) {
+            // pass can't be chosen, also an action that will be played by opposing player can't be chosen
+            if (action == VirtualActionToAction(kVirtualPass, boardSize) ||
+                std::find(vec.begin(), vec.end(), action) != vec.end())
+                continue;
+
+            newState->ApplyAction(action);
+            if (newState->board_.getStoneCount()[0] == currStoneCount[0] &&
+                newState->board_.getStoneCount()[1] == currStoneCount[1])
+            { //random move was applied correctly, no captures were made
+                newState->ApplyAction(VirtualActionToAction(kVirtualPass, boardSize));
+                //std::cout << "Added to observation " << ActionToString(c, action) << "\n";
+                break;
+            } else {
+                //std::cout << "random action" << ActionToString(c, action) << " was unacceptable\n";
+                newState->UndoAction(-1, -1);
```

Comment:
Avoid using "magic" numbers like -1 whenever possible, and use the right way to call the function -- the last player and last action, in case someone update the UndoAction() in the future to not use the rollouts.

Reply:
I am sorry for this, as i was aware that the implementation in this game does not take in donsideration these values, i just typed in some numbers, not considering possible update of the implementation.